### ISSUE TITLE
volk_8u_x4_conv_k7_r2_8u generic kernel inconsistent with SSE implementation

### ISSUE THREAD

**Issue by Aang23** [gnuradio/volk#473]
Hi!
I recently ported one of my projects over to Android, building VOLK via the NDK (I had to patch a few things to make it build, I will open another issue on the topic later). About everything works as expected.

The only thing that did not work was volk_8u_x4_conv_k7_r2_8u (generic kernel). I initially thought this was an issue on my side with my unusual environment, so I pulled out a Raspberry Pi and was able to reproduce the issue reliably using volk_8u_x4_conv_k7_r2_8u as a streaming decoder (similarly to GNU Radio's CC Decoder). No issue is observed with another implementation such as libcorrect's.

The behavior observed is volk_8u_x4_conv_k7_r2_8u seemingly working normally, but outputting 0s instead.

**Comment by odrisci** [comment]
I must confess that I know nothing about this code, but [this line](https://github.com/gnuradio/volk/blob/79a111ba600c86ae937830970b00ebe0c711df5a/kernels/volk/volk_8u_x4_conv_k7_r2_8u.h#L118) and the one following it look suspicious to me. It seems to me that the difference will always be non negative, so the decisions will always be 1 except when the differences are precisely 0.

**Comment by Aang23** [comment]
> I must confess that I know nothing about this code, but [this line](https://github.com/gnuradio/volk/blob/79a111ba600c86ae937830970b00ebe0c711df5a/kernels/volk/volk_8u_x4_conv_k7_r2_8u.h#L118) and the one following it look suspicious to me. It seems to me that the difference will always be non negative, so the decisions will always be 1 except when the differences are precisely 0.

My understanding of this isn't the best either, but I do not see anything obviously wrong with this implementation compared to others that definitely do work. I'm honestly very surprised the generic kernel does not work, even more that it works perfectly fine on x86 and not arm. 

This kinda kills half of my usecase as I am using Volk's implementation the most. Guess I will have to debug on my end in the meantime.

**Comment by Aang23** [comment]
Actually, I was partially wrong... It does not 0s but garbage data. A lower BER gets it down to more 0s. Weird behavior...
Here it is visualized in a bitstream viewer.
![Screenshot from 2021-05-10 19-40-38](https://user-images.githubusercontent.com/23421917/117705439-90b95d80-b1bb-11eb-95c9-2bc0748e69bd.png)

**Comment by jdemel** [comment]
I'd suggest to test this kernel with a high SNR, or only very (or no) bit errors.
Also how exactly do you test this?

**Comment by Aang23** [comment]
I tested it in several different ways : 
- Feeding the r=1/2, k=7, G1=-79, G2=-109 encoded CCSDS sync word (0xfca2b63db00d9794), converted to 0 to 255 uint8_t values. This was successful. After this, I modified a few bits (using 0xfca2b63db11d9794), which should be successfully decoded to 0x1acffc1d . Works fine on x86, but on my 3 ARM machines I tested on, results were very inconsistent with the same decode giving different results on each run (all variables in my test program were initialized to 0 or some other value).
- Directly iin one my actual programs (using METEOR-M 2 LRPT Data and JPSS-1 HRD data), and comparing each processing step on several ARM machines and x86. This allowed me to find a few mistakes on my side, to now end with an identical input to the viterbi decoder, but different outputs. (correction on lower SNRs being about none)
- Generating known good test data (generated from 10MB of /dev/random), fed into the decoder on once again several machines. This did work fine with minor differences in the output.

In all cases, VOLK was compiled with GCC 10.2.0 in debug mode (-O0), as well as the test program. Partab, metrics, etc were generated using GNU Radio's CC Decoder's code (which is as far as I know the only sample implementation I could find using VOLK) and the generic kernel.

In conclusion, in all cases, volk_8u_x4_conv_k7_r2_8u produces inferior results on all ARM platforms I could test. (Raspberry Pi 2, my Oneplus 6T (running a linux container) and a RK3318 board I had laying around)

Edit : Issue renamed as it does somewhat work, even if it is not quite usable.
Edit 2 : Just thought I'd add the output on x86 is as expected and consistent.

**Comment by Aang23** [comment]
Actually, I just realized I was using the SSE kernel on my x86 machine... Sorry, my bad. 
Then, this means the generic kernel is in fact the issue here. Identical issues as seen on ARM are observed.

Edit : To show the difference, here is the decode produced using the spiral kernel https://ibb.co/dpVSFzz, and with generic https://ibb.co/2PJ6Ckb.

**Comment by jdemel** [comment]
Thanks for debugging this issue further.
This particular kernel has several issues. It fails for AVX2, it produces inconsistent output on different platforms, and it is very hard to read and understand the code, thus it is terribly difficult to maintain it.
I'd really like to see this issue fixed though.

**Comment by Aang23** [comment]
You're welcome!
Indeed. I initially thought only the AVX2 kernel had issues, but that turned out to be very much wrong once I was on a platform with only Generic available.

Definitely not a kernel I'm really able to help debugging more than that, I do not have in-depth knowledge of that algorithm so understanding and debugging this won't be very easy... Will give it a go whenever I can, but I'm not hoping for much if any success.

I would guess @trondeau would be the one to ask about this issue, as he wrote those kernels looking through the commit history.

**Comment by michaelld** [comment]
Getting Tom to comment will be challenging. We're pretty much left to hope that someone understands the algorithm well enough to make useful commentary. I'm sure I could work it out if I had time, since I've written convolutional trellis decoding implementations before ... many years ago ... but I really don't have time. Maybe someone else does?

IMHO it's worth trying tweaking [those 2 notes lines](https://github.com/gnuradio/volk/blob/79a111ba600c86ae937830970b00ebe0c711df5a/kernels/volk/volk_8u_x4_conv_k7_r2_8u.h#L118-L119) to read `((int)m0)-((int)m1)` etc ... they do look suspicious, and IIRC for bit determination we're basically subtracting the log likelihood and then doing a binary determination (positive or negative, with 0 going arbitrarily but consistently one way or the other). The current code will always be either > 0 or 0, which I don't think follows the algorithm. Worth a try.

If we have any kernel that is broken & nobody understands & we can't get it fixed, then IMHO it should be removed -- either completely removed from the code, or `#if 0` ... `#endif` to disable that section of code. I'm getting the feeling that this kernel -- in part or in whole -- falls into this category. I have no issues with kernels the implement a complicated algorithm ... but they need to work & reliably!

**Comment by michaelld** [comment]
@Aang23 do you have some test code that you can pass on that we can use to do debugging? Asking since I'm guessing you do have such code & that will be a lot faster than us trying to create our own! Thx!

**Comment by Aang23** [comment]
> @Aang23 do you have some test code that you can pass on that we can use to do debugging? Asking since I'm guessing you do have such code & that will be a lot faster than us trying to create our own! Thx!

Oh sorry, I had entirely missed your reply. I will cleanup my test code and post it a bit later.

After discussing with @jdemel on Matrix, it seems like I wasn't clear enough about the exact issue.
The current generic kernel performs just fine given error-free data, but has very poor error-correction performances, hence why it probably did not show up in QA earlier.
Feeding the kernel some slightly corrupted data (like every real-world application would likely have) will result in no FEC gain at all, hence why I initially noticed the issue in my usecase.

https://github.com/gnuradio/volk/issues/473#issuecomment-839957073

Here with those images my intent was to show how the generic kernel only decoded valid data at peak SNR, when other FEC layers could take over, while in contrast the spiral kernel performed much better.

**Comment by noc0lour** [comment]
Running some CC Enc & Dec with error-free data in GNU Radio results in garbage output, so not sure if it even works at all. It worked with GNU Radio 3.7 & corresponding VOLK, but doesn't anymore..
